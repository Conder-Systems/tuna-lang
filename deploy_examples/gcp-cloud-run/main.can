{
  "DEPLOYMENT_NAME": "local-run",
  "STORES": "{}",
  "PROCEDURES": "{\"fibonacci\":[{\"kind\":\"assertHeapLen\",\"data\":3},{\"kind\":\"enforceSchemaInstanceOnHeap\",\"data\":[0,{\"kind\":\"Any\"}]},{\"kind\":\"conditonallySkipXops\",\"data\":1},{\"kind\":\"raiseError\",\"data\":\"invalid input\"},{\"kind\":\"enforceSchemaInstanceOnHeap\",\"data\":[1,{\"kind\":\"Any\"}]},{\"kind\":\"conditonallySkipXops\",\"data\":1},{\"kind\":\"raiseError\",\"data\":\"invalid input\"},{\"kind\":\"enforceSchemaInstanceOnHeap\",\"data\":[2,{\"kind\":\"Any\"}]},{\"kind\":\"conditonallySkipXops\",\"data\":1},{\"kind\":\"raiseError\",\"data\":\"invalid input\"},{\"kind\":\"copyFromHeap\",\"data\":2},{\"kind\":\"instantiate\",\"data\":0},{\"kind\":\"equal\"},{\"kind\":\"negatePrev\"},{\"kind\":\"conditonallySkipXops\",\"data\":3},{\"kind\":\"copyFromHeap\",\"data\":1},{\"kind\":\"returnStackTop\"},{\"kind\":\"offsetOpCursor\",\"data\":[1,true]},{\"kind\":\"noop\"},{\"kind\":\"copyFromHeap\",\"data\":1},{\"kind\":\"copyFromHeap\",\"data\":0},{\"kind\":\"copyFromHeap\",\"data\":1},{\"kind\":\"plus\"},{\"kind\":\"copyFromHeap\",\"data\":2},{\"kind\":\"instantiate\",\"data\":1},{\"kind\":\"nMinus\"},{\"kind\":\"invoke\",\"data\":[\"fibonacci\",3]},{\"kind\":\"returnStackTop\"}],\"try_fib\":[{\"kind\":\"assertHeapLen\",\"data\":1},{\"kind\":\"enforceSchemaInstanceOnHeap\",\"data\":[0,{\"kind\":\"Any\"}]},{\"kind\":\"conditonallySkipXops\",\"data\":1},{\"kind\":\"raiseError\",\"data\":\"invalid input\"},{\"kind\":\"copyFromHeap\",\"data\":0},{\"kind\":\"instantiate\",\"data\":10},{\"kind\":\"lesseq\"},{\"kind\":\"negatePrev\"},{\"kind\":\"negatePrev\"},{\"kind\":\"conditonallySkipXops\",\"data\":3},{\"kind\":\"instantiate\",\"data\":\"too much fib no thanks\"},{\"kind\":\"returnStackTop\"},{\"kind\":\"offsetOpCursor\",\"data\":[1,true]},{\"kind\":\"noop\"},{\"kind\":\"instantiate\",\"data\":1},{\"kind\":\"instantiate\",\"data\":1},{\"kind\":\"copyFromHeap\",\"data\":0},{\"kind\":\"invoke\",\"data\":[\"fibonacci\",3]},{\"kind\":\"returnStackTop\"}]}",
  "PRIVATE_PROCEDURES": "[\"fibonacci\"]",
  "SCHEMAS": "{}"
}