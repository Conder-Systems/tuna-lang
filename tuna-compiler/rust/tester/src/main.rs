
use tuna_interpreter::{conduit_byte_code_interpreter, Globals};
use tuna_interpreter::ops::{Op};
use std::collections::HashMap;
use serde::{Deserialize};
use std::fs;

#[derive(Deserialize)]
struct Main {
    main: Vec<Op>,
}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    let main: Main = serde_json::from_str(
        std::str::from_utf8(
            fs::read(&"main.can")?.as_slice()
        ).unwrap())?;

    
    Ok(())
}
