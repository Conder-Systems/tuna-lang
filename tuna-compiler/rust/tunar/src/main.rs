
use tuna_interpreter::{conduit_byte_code_interpreter, Globals, Context};
use tuna_interpreter::ops::{Op};
use tuna_interpreter::schemas::{Schema};
use std::collections::HashMap;
use serde::{Deserialize};
use std::fs;
use crypto::ed25519;
use rand;
use rand_core::RngCore;
use ts_rs::{TS, export};
use std::any::TypeId;

#[derive(Deserialize, TS)]
struct Runnable {
    main: Vec<Op>,
    lookups: Lookups
}

#[derive(Deserialize)]
struct Lookups {
    schemas: HashMap<String, Schema>, 
    functions: HashMap<String, Vec<Op>>
}

impl TS for Lookups {
    fn name() -> String {
        return r#"{
            schemas: Record<string, Schema>,
            functions: Record<string, Op[]>
        }
        "#.to_string();
    }

    fn dependencies() -> Vec<(TypeId, String)>{
        return vec![]
    }

    fn transparent() -> bool {
        return false
    }

    fn inline(_indent: usize) -> String {
        return r#"{
            schemas: Record<string, Schema>,
            functions: Record<string, Op[]>
        }
        "#.to_string();
    }

}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    let state: Runnable = serde_json::from_str(
        std::str::from_utf8(
            fs::read(&"main.can")?.as_slice()
        ).unwrap())?;

    let stores = HashMap::new();
    let mut key = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut key);
    let (priv_key, pub_key) = ed25519::keypair(&key);

    let res = conduit_byte_code_interpreter(
        Context::new(&state.main, vec![]),
        &Globals {
            schemas: &state.lookups.schemas,
            fns: &state.lookups.functions,
            stores: &stores,
            private_key: &priv_key,
            public_key: &pub_key
        }
    ).await;

    match res {
        Ok(_) => Ok(()),
        Err(e) => panic!("Failed running with: {}", e)
    }
}

export! {
    Runnable => "runnable.ts"
}