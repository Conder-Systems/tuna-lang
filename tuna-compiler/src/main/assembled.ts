import {
    Manifest, 
    StrongServerEnv,
    ServerEnv,
    functionsToOps,
} from './backend/index'
import * as ed from 'noble-ed25519'

import {Parser} from './frontend/parser'
import {semantify, PrivateFuncs, Schemas } from './frontend/semantics'

export function TUNA_TO_MANIFEST(s: string): Manifest & PrivateFuncs & Schemas {
    const p = new Parser(s).parse()
    return semantify(p, false)
}
type Keys = "PRIVATE_KEY" | "PUBLIC_KEY"
type Keyless = Omit<StrongServerEnv, Keys>

export function STRINGIFY_ENV(env: Keyless): Omit<ServerEnv, Keys> {
    
    const string_env: Partial<ServerEnv> = {};
    for (const key in env) {
        //@ts-ignore
        string_env[key] = typeof env[key] === "string" ? env[key] : JSON.stringify(env[key]);
    }
    return string_env as ServerEnv
}

export function TUNA_TO_ENV(s: string): Keyless {
    const man = TUNA_TO_MANIFEST(s)
    const PROCEDURES = functionsToOps(man.funcs)
    const STORES: StrongServerEnv["STORES"] = {}
    man.globals.forEach((v, k) => STORES[k] = {kind: "Any", data: null})
    return {
        DEPLOYMENT_NAME: "local-run",
        STORES,
        PROCEDURES,
        PRIVATE_PROCEDURES: [...man.privateFuncs.values()],
        SCHEMAS: man.schemas
    }
    
}


export const TUNA_LOCAL_COMPILER = (s: string) => STRINGIFY_ENV(TUNA_TO_ENV(s))