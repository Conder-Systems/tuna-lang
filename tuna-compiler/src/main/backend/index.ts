
export * from './abstract/IR'
export * from './abstract/function'
export * from './abstract/globals/mongo'
export * from './abstract/mongo_logic/main'
export {Transform, Transformer, Compiler}from './abstract/compilers'
export * from './ops/index'