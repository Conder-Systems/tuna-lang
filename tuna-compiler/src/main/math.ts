import { PickNode, ValueNode } from "conder_core";


export type Sign = PickNode<"Math">["sign"]
type MathNode = PickNode<"Math">["left"]
export type MathExpression = {
    then(sign: Sign, value: MathNode): MathExpression
    build(): ValueNode
}
type MathTup = [Sign, MathNode]
type MathInstr = MathTup[]

export class Ordering implements MathExpression {
    //always len(values) == len(signs) + 1
    private readonly first: MathNode
    private readonly instrs: MathInstr[]
    private state: "+-" | "*" | "/" | "uninitialized" = "uninitialized"
    constructor(first: MathNode) {
        this.first = first
        this.instrs = []
    }

    private startMultiplication(state: Ordering["state"], sign: Sign, value: MathNode) {
        const last = this.instrs.pop()
        
        const steal = last.pop()
        if (last.length !== 0) {
            this.instrs.push(last)
        }
        this.instrs.push([steal, [sign, value]])
        this.state = state
    }
    pushToGroup(sign: Sign, value: MathNode) {
        const last = this.instrs.pop()
        last.push([sign, value])
        this.instrs.push(last)
    }
    then(sign: Sign, value: MathNode): MathExpression {

        switch (sign) {
            case "+":
            case "-":
                this.state = "+-"
                this.instrs.push([[sign, value]])
                break
            case "*":
            case "/":
                switch (this.state) {
                    case "+-":
                        this.startMultiplication(sign, sign, value)
                        break
                    case "uninitialized":
                        this.state = sign
                        this.instrs.push([[sign, value]])
                        break
                    case "*":
                    case "/":
                        this.pushToGroup(sign === this.state ? "*" : "/", value)
                }
            
        }
        
         
        return this
    }

    build(): ValueNode {
        let top = this.first
        // console.log(JSON.stringify(this.instrs, null, 2))
        const grouped = this.group()
        // console.log(JSON.stringify(grouped, null, 2))
        grouped.forEach(([sign, right]) => {
            top = {
                kind: "Math",
                left: top,
                right,
                sign
            }
        })
        return top
    }

    private group(): MathTup[] {
        
        const tups: MathTup[] = []
        for (let i = 0; i < this.instrs.length; i++) {
            
            const group = this.instrs[i];
            if (group.length === 0) {
                continue
            }
            const lead_tup = group[0]
            const group_sign = lead_tup[0]
            let agg = lead_tup[1]

            for (let j = 1; j < group.length; j++) {
                const [sign, right] = group[j]
                agg = {
                    kind: "Math",
                    sign,
                    left: agg,
                    right
                }
            }
            tups.push([group_sign, agg])
        }
        return tups
    }
}