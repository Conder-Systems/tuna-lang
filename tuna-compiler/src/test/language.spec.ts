import { OPSIFY_MANIFEST, Transformer } from 'conder_core';
import { TUNA_TO_MANIFEST } from '../main/assembled';

describe("language", () => {

    function tunaTest(maybeSucceed: "succeed" | "fail", code: string): jest.ProvidesCallback {
        return (cb) => {
            if (maybeSucceed === "succeed") {
                const ops = TUNA_TO_MANIFEST.then(new Transformer(i => {
                    expect(i).toMatchSnapshot("intermediate representation")
                    return i
                })).then(OPSIFY_MANIFEST).run(code)

            } else {
                expect(() => TUNA_TO_MANIFEST.run(code)).toThrowErrorMatchingSnapshot()
            }
            cb()
        }
    }

    it("should allow a global object", tunaTest("succeed", `const obj = {}`))
        

    it("should allow many global objects",
        tunaTest("succeed", `
        const obj1 = {}
        const obj2 = {}
        `)
    )

    it("should allow empty pub funcs", 
    tunaTest("succeed", `pub func doSomething() {}`))

    it("should allow a fixed number of args in functions",
    tunaTest("succeed", `pub func argy(a, b, c) {

    }`)
    )

    it("should allow return statements within functions",
    tunaTest("succeed", `pub func returny() {
        return
    }`))

    it("should allow setting of keys on a global object", 
    
        tunaTest("succeed",
        `
        const gg = {}
        pub func fff(a) {
            gg.abc = a
            gg[a] = a
            gg['abc'] = a
        }
        `)
    )


    it("should allow getting of nested keys",
    
        tunaTest("succeed",
        `
        const gg = {}
        pub func fff(a) {
            return gg[a].field
        }
        `)
    )

    it('should allow bools, numbers, and strings', tunaTest("succeed", `
    
    pub func fff(a) {
        true
        false
        12
        -12.12
        'hello world'
        {}
    }
    
    `))
    
    it('can declare temp variables', tunaTest("succeed", `
    
    pub func fff(a) {
        const b = true
        let c = false
        const d = a[b]
    }
    
    `))

    it('can overwrite inputs', tunaTest("succeed", `
    
    pub func fff(a) {
        a = false
    }
    `))

    it('cannot have duplicate variables', tunaTest("fail", `
    
    pub func fff(a) {
        const a = true
    }
    `))

    it('cannot have a variable with the same name as a function', tunaTest("fail", `
    
    pub func fff(a) {
        const fff = 12
    }
    `))

    it("does not allow overwriting constants", tunaTest("fail",`
    pub func fff(a) {
        const b = a
        b = 42
    }

    `))
    
    it('only allows global constants', tunaTest("fail", `    
    let someVar = {}
    `))

    it('globals must be empty objects', tunaTest("fail", `
    const someVar = false
    `))

    it("allows users to call keys() on globals and locals", tunaTest("succeed",
    `
    const g = {}
    pub func f(a) {
        const b = a.keys()
        const gk = g.keys()
    }
    `
    ))

    it("allows keys on nested object", tunaTest("succeed", 
    `
    pub func f(a) {
        return a['b'].cdef.keys()
    }
    `
    ))

    //Blocked by https://github.com/Conder-Systems/conder/issues/66
    it.skip("allows indexing into keys results", tunaTest("succeed",
    `
    pub func f(a) {
        return a.keys()[0]
    }
    `))

    it("allows deleting of keys in objects", tunaTest("succeed",
    `
    pub func f(a) {
        delete(a.b)
    }
    `
    ))

    it("shouldn't allow deleting of whole variables", tunaTest("fail",
    `
    pub func f(b0) {
        delete(b0)
    }
    `
    ))

    // #dontlike 
    // delete should work on arrays like it does on objects
    it("allows deleting of array fields even though it produces a runtime error", tunaTest("succeed",
    `
    pub func f(b) {
        delete(b[0])
    }
    `
    ))

    it("allows for loops", tunaTest(
        "succeed",
        `
        pub func loop(arr) {
            for row in arr {
                return row
            }
        }
        `    
    ))

    it("ensure variables declared in for loop are cleaned up", tunaTest(
        "succeed",
        `
        const g = {}
        pub func loop() {
            for row in g.array {
                let some_scoped_var = row
            }
            const should_be_at_index_0 = g.array
            return should_be_at_index_0 
        }
        `
    ))

    it("ifs", tunaTest(
        "succeed",
        `
        pub func maybe(a) {
            if a {
                return a
            }
        }
        `
    ))
})