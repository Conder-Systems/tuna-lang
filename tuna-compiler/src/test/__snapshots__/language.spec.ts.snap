// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`language allows deleting of array fields even though it produces a runtime error: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "f" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Update",
          "level": Array [
            Object {
              "kind": "Int",
              "value": 0,
            },
          ],
          "operation": Object {
            "kind": "DeleteField",
          },
          "root": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language allows deleting of keys in objects: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "f" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Update",
          "level": Array [
            Object {
              "kind": "String",
              "value": "b",
            },
          ],
          "operation": Object {
            "kind": "DeleteField",
          },
          "root": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language allows for loops: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "loop" => FunctionDescription {
      "computation": Array [
        Object {
          "do": Array [
            Object {
              "kind": "Return",
              "value": Object {
                "index": 1,
                "kind": "Saved",
              },
            },
          ],
          "kind": "ArrayForEach",
          "target": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language allows keys on nested object: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "f" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Return",
          "value": Object {
            "kind": "Keys",
            "target": Object {
              "kind": "Selection",
              "level": Array [
                Object {
                  "kind": "String",
                  "value": "b",
                },
                Object {
                  "kind": "String",
                  "value": "cdef",
                },
              ],
              "root": Object {
                "index": 0,
                "kind": "Saved",
              },
            },
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language allows users to call keys() on globals and locals: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "f" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Keys",
            "target": Object {
              "index": 0,
              "kind": "Saved",
            },
          },
        },
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Keys",
            "target": Object {
              "kind": "GlobalObject",
              "name": "g",
            },
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {
    "g" => Object {
      "kind": "glob",
      "name": "g",
    },
  },
}
`;

exports[`language can declare temp variables: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "fff" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Bool",
            "value": true,
          },
        },
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Bool",
            "value": false,
          },
        },
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Selection",
            "level": Array [
              Object {
                "index": 1,
                "kind": "Saved",
              },
            ],
            "root": Object {
              "index": 0,
              "kind": "Saved",
            },
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language can overwrite inputs: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "fff" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Update",
          "level": Array [],
          "operation": Object {
            "kind": "Bool",
            "value": false,
          },
          "root": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language cannot have a variable with the same name as a function 1`] = `
"In function fff: 
	The name fff is initialized to multiple variables/functions."
`;

exports[`language cannot have duplicate variables 1`] = `
"In function fff: 
	The name a is initialized to multiple variables/functions."
`;

exports[`language does not allow overwriting constants 1`] = `
"In function fff: 
	Attempting to overwrite constant variable b"
`;

exports[`language ensure variables declared in for loop are cleaned up: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "loop" => FunctionDescription {
      "computation": Array [
        Object {
          "do": Array [
            Object {
              "kind": "Save",
              "value": Object {
                "index": 0,
                "kind": "Saved",
              },
            },
          ],
          "kind": "ArrayForEach",
          "target": Object {
            "kind": "Selection",
            "level": Array [
              Object {
                "kind": "String",
                "value": "array",
              },
            ],
            "root": Object {
              "kind": "GlobalObject",
              "name": "g",
            },
          },
        },
        Object {
          "kind": "Save",
          "value": Object {
            "kind": "Selection",
            "level": Array [
              Object {
                "kind": "String",
                "value": "array",
              },
            ],
            "root": Object {
              "kind": "GlobalObject",
              "name": "g",
            },
          },
        },
        Object {
          "kind": "Return",
          "value": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [],
    },
  },
  "globals": Map {
    "g" => Object {
      "kind": "glob",
      "name": "g",
    },
  },
}
`;

exports[`language globals must be empty objects 1`] = `"Global someVar must be initialized as empty object"`;

exports[`language ifs allows any number of else ifs: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "kind": "String",
                    "value": "a",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "cond": Object {
                "kind": "Selection",
                "level": Array [
                  Object {
                    "kind": "String",
                    "value": "b",
                  },
                ],
                "root": Object {
                  "index": 0,
                  "kind": "Saved",
                },
              },
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "kind": "String",
                    "value": "b",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "cond": Object {
                "kind": "Selection",
                "level": Array [
                  Object {
                    "kind": "String",
                    "value": "c",
                  },
                ],
                "root": Object {
                  "index": 0,
                  "kind": "Saved",
                },
              },
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "kind": "String",
                    "value": "c",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "kind": "String",
                    "value": "d",
                  },
                },
              ],
              "kind": "Else",
            },
          ],
          "kind": "If",
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs allows elses: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [],
              "kind": "Conditional",
            },
            Object {
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Else",
            },
          ],
          "kind": "If",
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs can 'not' values: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "negate" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Return",
          "value": Object {
            "kind": "Comparison",
            "left": Object {
              "index": 0,
              "kind": "Saved",
            },
            "right": Object {
              "kind": "Bool",
              "value": false,
            },
            "sign": "==",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs cleans up variables across else ifs: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "cond": Object {
                "kind": "Selection",
                "level": Array [
                  Object {
                    "kind": "String",
                    "value": "b",
                  },
                ],
                "root": Object {
                  "index": 0,
                  "kind": "Saved",
                },
              },
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "cond": Object {
                "kind": "Selection",
                "level": Array [
                  Object {
                    "kind": "String",
                    "value": "c",
                  },
                ],
                "root": Object {
                  "index": 0,
                  "kind": "Saved",
                },
              },
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Else",
            },
          ],
          "kind": "If",
        },
        Object {
          "kind": "Save",
          "value": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs cleans up variables in ifs: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
          ],
          "kind": "If",
        },
        Object {
          "kind": "Save",
          "value": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
        Object {
          "kind": "Return",
          "value": Object {
            "index": 1,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs else scope is cleaned up after each block: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
            Object {
              "do": Array [
                Object {
                  "kind": "Save",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Else",
            },
          ],
          "kind": "If",
        },
        Object {
          "kind": "Save",
          "value": Object {
            "index": 0,
            "kind": "Saved",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language ifs single simple if: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "maybe" => FunctionDescription {
      "computation": Array [
        Object {
          "conditionally": Array [
            Object {
              "cond": Object {
                "index": 0,
                "kind": "Saved",
              },
              "do": Array [
                Object {
                  "kind": "Return",
                  "value": Object {
                    "index": 0,
                    "kind": "Saved",
                  },
                },
              ],
              "kind": "Conditional",
            },
          ],
          "kind": "If",
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language only allows global constants 1`] = `"Global variable someVar must be const"`;

exports[`language should allow a fixed number of args in functions: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "argy" => FunctionDescription {
      "computation": Array [],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
        Object {
          "data": undefined,
          "kind": "Any",
        },
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language should allow a global object: intermediate representation 1`] = `
Object {
  "funcs": Map {},
  "globals": Map {
    "obj" => Object {
      "kind": "glob",
      "name": "obj",
    },
  },
}
`;

exports[`language should allow bools, numbers, and strings: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "fff" => FunctionDescription {
      "computation": Array [],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {},
}
`;

exports[`language should allow empty pub funcs: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "doSomething" => FunctionDescription {
      "computation": Array [],
      "input": Array [],
    },
  },
  "globals": Map {},
}
`;

exports[`language should allow getting of nested keys: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "fff" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Return",
          "value": Object {
            "kind": "Selection",
            "level": Array [
              Object {
                "index": 0,
                "kind": "Saved",
              },
              Object {
                "kind": "String",
                "value": "field",
              },
            ],
            "root": Object {
              "kind": "GlobalObject",
              "name": "gg",
            },
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {
    "gg" => Object {
      "kind": "glob",
      "name": "gg",
    },
  },
}
`;

exports[`language should allow many global objects: intermediate representation 1`] = `
Object {
  "funcs": Map {},
  "globals": Map {
    "obj1" => Object {
      "kind": "glob",
      "name": "obj1",
    },
    "obj2" => Object {
      "kind": "glob",
      "name": "obj2",
    },
  },
}
`;

exports[`language should allow return statements within functions: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "returny" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Return",
          "value": undefined,
        },
      ],
      "input": Array [],
    },
  },
  "globals": Map {},
}
`;

exports[`language should allow setting of keys on a global object: intermediate representation 1`] = `
Object {
  "funcs": Map {
    "fff" => FunctionDescription {
      "computation": Array [
        Object {
          "kind": "Update",
          "level": Array [
            Object {
              "kind": "String",
              "value": "abc",
            },
          ],
          "operation": Object {
            "index": 0,
            "kind": "Saved",
          },
          "root": Object {
            "kind": "GlobalObject",
            "name": "gg",
          },
        },
        Object {
          "kind": "Update",
          "level": Array [
            Object {
              "index": 0,
              "kind": "Saved",
            },
          ],
          "operation": Object {
            "index": 0,
            "kind": "Saved",
          },
          "root": Object {
            "kind": "GlobalObject",
            "name": "gg",
          },
        },
        Object {
          "kind": "Update",
          "level": Array [
            Object {
              "kind": "String",
              "value": "abc",
            },
          ],
          "operation": Object {
            "index": 0,
            "kind": "Saved",
          },
          "root": Object {
            "kind": "GlobalObject",
            "name": "gg",
          },
        },
      ],
      "input": Array [
        Object {
          "data": undefined,
          "kind": "Any",
        },
      ],
    },
  },
  "globals": Map {
    "gg" => Object {
      "kind": "glob",
      "name": "gg",
    },
  },
}
`;

exports[`language shouldn't allow deleting of whole variables 1`] = `
"In function f: 
	Delete cannot be applied to whole variables"
`;
